bee-transaction/src/atomic/payload/signed_transaction/mod.rs

```diff
@@ -0,0 +1,333 @@
+            match i {
+                Input::UTXO(u) => {
+                    // Transaction Output Index must be 0 ≤ x < 127
+                    match u.index() {
+                        0..=126 => (),
+                        _ => return Err(Error::CountError),
+                    }
+
+                    // Every combination of Transaction ID + Transaction Output Index must be unique in the inputs set.
+                    if !insert_combination(&mut combination, u) {
+                        return Err(Error::DuplicateError);
+                    }
+                }
+            }
+        }
+
+        // Inputs must be in lexicographical order of their serialized form.
+        if !is_sorted(transaction.inputs.iter()) {
+            return Err(Error::OrderError);
+        }
+
+        // Output validation
+        // Outputs Count must be 0 < x < 127
+        match transaction.outputs.len() {
+            1..=126 => (),
+            _ => return Err(Error::CountError),
+        }
+
+        // At least one output must be specified
+        if transaction.outputs.is_empty() {
+            return Err(Error::EmptyError);
+        }
+
+        let mut total = 0;
+        let mut combination = Vec::new();
+        for i in transaction.outputs.iter() {
+            // Output Type must be 0, denoting a SigLockedSingleDeposit.
+            match i {
+                output::Output::SigLockedSingleDeposit(u) => {
+                    // Address Type must either be 0 or 1, denoting a WOTS- or Ed25519 address.
+
+                    // If Address is of type WOTS address, its bytes must be valid T5B1 bytes.
+
+                    // The Address must be unique in the set of SigLockedSingleDeposits
+                    if !insert_combination(&mut combination, u.address()) {
+                        return Err(Error::DuplicateError);
+                    }
+
+                    // Amount must be > 0
+                    let amount = u.amount().get();
+                    if amount == 0 {
+                        return Err(Error::AmountError);
+                    }
+
+                    total += amount;
+                }
+            }
+        }
+
+        // Outputs must be in lexicographical order by their serialized form
+        if !is_sorted(transaction.outputs.iter()) {
+            return Err(Error::OrderError);
+        }
+
+        // Accumulated output balance must not exceed the total supply of tokens 2'779'530'283'277'761
+        if total > 2779530283277761 {
+            return Err(Error::AmountError);
+        }
+
+        // Payload Length must be 0 (to indicate that there's no payload) or be valid for the specified payload type.
+        // Payload Type must be one of the supported payload types if Payload Length is not 0.
+
+        // Unlock Blocks validation
+        // Unlock Blocks Count must match the amount of inputs. Must be 0 < x < 127.
+        match self.unlock_block_count {
```

Comment:
Given that there are a lot of bounds checks maybe factoring this out into something like
```rust
fn check_bound<E>(low: usize, high: usize, value: usize, err: impl Fn() -> E) -> Result<(), E> { ... }
```
would be a good idea.

Reply:
I realized we could use `contains` method from `Range` which should also do the work.